packages/supabase/src/index.ts

```diff
@@ -0,0 +1,176 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import type {
+  Adapter,
+  AdapterSession,
+  AdapterUser,
+  VerificationToken,
+} from "next-auth/adapters"
+import type { SupabaseClient } from "@supabase/supabase-js"
+import type { Account } from "next-auth"
+
+export const formatter = {
+  /** Takes a Fauna object and returns a plain old JavaScript object */
+  from<T = Record<string, unknown>>(object: Record<string, any>): T {
+    const newObject: Record<string, unknown> = {}
+    for (const key in object) {
+      const value = object[key]
+      if (key === "id" || key === "userId") {
+        newObject[key] = value.toString()
+      } else if (typeof value === "string" && Date.parse(value)) {
+        newObject[key] = new Date(value)
+      } else {
+        newObject[key] = value
+      }
+    }
+    return newObject as T
+  },
+  to(object: Record<string, any>) {
+    const newObject: Record<string, unknown> = {}
+    for (const key in object) {
+      const value = object[key]
+      if (key === "id" || key === "userId") {
+        newObject[key] = parseInt(value, 10)
+      } else {
+        newObject[key] = value
+      }
+    }
+    return newObject
+  },
+}
+
+export function SupabaseAdapter(client: SupabaseClient): Adapter {
+  const { from, to } = formatter
+  return {
+    async createUser(data) {
+      const result = await client
+        .from<AdapterUser>("users")
+        .insert(to(data))
+        .single()
+      return from<AdapterUser>(result.data!)
+    },
+    async getUser(id) {
+      const result = await client
+        .from<AdapterUser>("users")
+        .select()
+        .eq("id", id)
+        .single()
+      if (!result.data) return null
+      return from<AdapterUser>(result.data)
+    },
+    async getUserByEmail(email) {
+      const result = await client
+        .from<AdapterUser>("users")
+        .select()
+        .eq("email", email)
+        .single()
+
+      if (!result.data) return null
+      return from<AdapterUser>(result.data)
+    },
+    async getUserByAccount({ provider, providerAccountId }) {
+      const account = await client
+        .from<Account>("accounts")
+        .select()
+        .eq("provider", provider)
+        .eq("providerAccountId", providerAccountId)
+        .single()
+
+      const userId = account.data?.userId
+      if (!userId) return null
+
+      return await this.getUser(userId)
```

Comment:
I've been testing this out. And seems like the reference to `this` is being lost. I have changed this line for `return adapter.getUser(userId)` and instead of returning the object, I have bonded it to a variable `adapter` and it works.


So I have something like:

```
const adapter = {
...
    async getUserByAccount({ provider, providerAccountId }) {
      const account = await client
        .from<Account>("accounts")
        .select()
        .eq("provider", provider)
        .eq("providerAccountId", providerAccountId)
        .single()

      const userId = account.data?.userId
      if (!userId) return null

      return adapter.getUser(userId)
},
...
}

return adapter
```
```

Reply:
this also seemed to work:

```
    async getUserByAccount({ provider, providerAccountId }) {
      const result = await client
        .from<Account>('accounts')
        .select('user: users(*)')
        .eq('provider', provider)
        .eq('providerAccountId', providerAccountId)
        .single()

      if (!result.data) return null

      return from<AdapterUser>(result.data.user)
    },
```

Reply:
@balazsorban44 can you have a look ? thanks